File: app/src/main/java/me/zhanghai/android/files/settings/StandardDirectoriesPreference.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.settings;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.widget.TextView;

import java.util.List;

import androidx.annotation.AttrRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;
import androidx.lifecycle.Observer;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;
import me.zhanghai.android.files.compat.ListFormatterCompat;
import me.zhanghai.android.files.navigation.StandardDirectoriesLiveData;
import me.zhanghai.android.files.navigation.StandardDirectory;
import me.zhanghai.java.functional.Functional;

public class StandardDirectoriesPreference extends Preference {

    private final Observer<List<StandardDirectory>> mObserver = this::onStandardDirectoriesChanged;

    @Nullable
    private CharSequence mSummaryEmpty;

    public StandardDirectoriesPreference(@NonNull Context context) {
        super(context);

        init();
    }

    public StandardDirectoriesPreference(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public StandardDirectoriesPreference(@NonNull Context context, @Nullable AttributeSet attrs,
                                         @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init();
    }

    public StandardDirectoriesPreference(@NonNull Context context, @Nullable AttributeSet attrs,
                                         @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        init();
    }

    private void init() {
        setPersistent(false);
        mSummaryEmpty = getSummary();
    }

    @Override
    public void onAttached() {
        super.onAttached();

        StandardDirectoriesLiveData.getInstance().observeForever(mObserver);
    }

    @Override
    public void onDetached() {
        super.onDetached();

        StandardDirectoriesLiveData.getInstance().removeObserver(mObserver);
    }

    private void onStandardDirectoriesChanged(
            @NonNull List<StandardDirectory> standardDirectories) {
        Context context = getContext();
        List<String> titles = Functional.map(Functional.filter(standardDirectories,
                StandardDirectory::isEnabled), standardDirectory -> standardDirectory.getTitle(
                context));
        CharSequence summary = !titles.isEmpty() ? ListFormatterCompat.format(titles)
                : mSummaryEmpty;
        setSummary(summary);
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder holder) {
        super.onBindViewHolder(holder);

        TextView summaryText = (TextView) holder.findViewById(android.R.id.summary);
        summaryText.setSingleLine(true);
        summaryText.setEllipsize(TextUtils.TruncateAt.END);
    }

    @Override
    protected void onClick() {
        Context context = getContext();
        context.startActivity(StandardDirectoriesActivity.newIntent(context));
    }
}

File: app/src/main/java/me/zhanghai/android/files/settings/StandardDirectoriesPreference.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.settings

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.widget.TextView
import androidx.annotation.AttrRes
import androidx.annotation.StyleRes
import androidx.lifecycle.Observer
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import me.zhanghai.android.files.compat.ListFormatterCompat
import me.zhanghai.android.files.navigation.StandardDirectoriesLiveData
import me.zhanghai.android.files.navigation.StandardDirectory
import me.zhanghai.android.files.util.createIntent

class StandardDirectoriesPreference : Preference {
    private val observer = Observer<List<StandardDirectory>> { onStandardDirectoriesChanged(it) }
    private var emptySummary = summary

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int) : super(
        context, attrs, defStyleAttr
    )

    constructor(
        context: Context,
        attrs: AttributeSet?,
        @AttrRes defStyleAttr: Int,
        @StyleRes defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        isPersistent = false
    }

    override fun onAttached() {
        super.onAttached()

        StandardDirectoriesLiveData.observeForever(observer)
    }

    override fun onDetached() {
        super.onDetached()

        StandardDirectoriesLiveData.removeObserver(observer)
    }

    private fun onStandardDirectoriesChanged(standardDirectories: List<StandardDirectory>) {
        val context = context
        val titles = standardDirectories.filter { it.isEnabled }.map { it.getTitle(context) }
        val summary = if (titles.isNotEmpty()) ListFormatterCompat.format(titles) else emptySummary
        setSummary(summary)
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)

        val summaryText = holder.findViewById(android.R.id.summary) as TextView
        summaryText.ellipsize = TextUtils.TruncateAt.END
        summaryText.isSingleLine = true
    }

    override fun onClick() {
        context.startActivity(StandardDirectoryListActivity::class.createIntent())
    }
}
